python/lsst/sims/utils/timeTransformations.py

```diff
@@ -0,0 +1,195 @@
+"""
+For definitions of all (or, at least, most of) the time systems handled in this file, see
+
+https://www.cv.nrao.edu/~rfisher/Ephemerides/times.html
+"""
+
+import numpy as np
+import warnings
+import palpy
+from lsst.sims.utils import Ut1MinusUtcData
+from lsst.sims.utils import TT_from_TAI_Data
+
+__all__ = ["taiFromUtc", "utcFromTai",
+           "dutFromUtc",
+           "ut1FromUtc", "utcFromUt1",
+           "dttFromUtc", "ttFromTai",
+           "tdbFromTt"]
+
+
+def taiFromUtc(utc):
+    """
+    Use the palpy method palDat to convert from Coordinated Universal Time
+    (UTC) to International Atomic Time (TAI)
+
+    @param [in] utc is the UTC time as an MJD
+
+    @param [out] TAI time as an MJD
+    """
+
+    dt = palpy.dat(utc) # returns TAI-UTC in seconds
+    return utc + dt/86400.0
+
+
+def utcFromTai(tai):
+    """
+    Use the palpy method palDat to convert from International Atomic Time
+    (TAI) to Coordinated Universal Time (UTC)
+
+    @param [in] tai is the TAI time as an MJD
+
+    @param [out] utc time as an MJD
+    """
+
+    # because the PALPY method only returns TAI-UTC as a function
+    # of UTC, we will have to construct an array of TAI and UTC
+    # values using taiFromUtc and interpolate along them
+
+    sec_to_day = 1.0/86400.0
+
+    dt_approx = palpy.dat(tai)*sec_to_day
+
+    if dt_approx == 0.0:
+        return tai
+
+    utc_arr = np.arange(tai - 1.0*dt_approx, tai+1.0*dt_approx, 1.0e-6)
+    tai_arr = np.array([taiFromUtc(utc) for utc in utc_arr])
+
+    return np.interp(tai, tai_arr, utc_arr)
+
+
+def dutFromUtc(utc):
+    """
+    Use data downloaded from
+
+    ftp://cddis.gsfc.nasa.gov/pub/products/iers/
+
+    to return UT1-UTC as a function of UTC
+
+    @param [in] UTC as an MJD
+
+    @param [out] UT1-UTC in seconds
+    """
+
+    if utc<Ut1MinusUtcData._mjd_arr[0] or utc>Ut1MinusUtcData._mjd_arr[-1]:
+        msg = "UTC = %e is outside of the the bounds " % utc \
+               + "for which we have UT1-UTC "\
+               + "data (%e <= utc <= %e)\n" % (Ut1MinusUtcData._mjd_arr[0], Ut1MinusUtcData._mjd_arr[-1]) \
+               + "We will return UT1-UTC = 0, for lack of a better idea"
+        warnings.warn(msg)
+
+        return 0.0
+
+    return np.interp(utc, Ut1MinusUtcData._mjd_arr, Ut1MinusUtcData._dut_arr)
```

Comment:
Okay - but I don't think that np.interp does the right thing here though. Example:

>>> mjd_dut
array([49353, 49718, 50083, 50448, 50813, 51178, 51543, 51908, 52273,
       52638, 53003, 53368, 53733, 54098, 54463, 54828, 55193, 55558])
>>> dut
array([ -1.00000000e-01,  -2.22044605e-16,   1.00000000e-01,
         2.00000000e-01,   3.00000000e-01,   4.00000000e-01,
         5.00000000e-01,   6.00000000e-01,   7.00000000e-01,
         8.00000000e-01,  -9.00000000e-01,  -8.00000000e-01,
        -7.00000000e-01,  -6.00000000e-01,  -5.00000000e-01,
        -4.00000000e-01,  -3.00000000e-01,  -2.00000000e-01])
>>> mjd
53000
>>> np.interp(mjd, mjd_dut, dut)
-0.8860273972602738

but really (because it's a discontinuous jump), a date just before 53003 should give a positive value.
>>> mjd_dut[9], dut[9]
(52638, 0.79999999999999971)
>>> mjd_dut[10], dut[10]
(53003, -0.90000000000000002)